### PR TITLE
Fix an issue with regard to finishing out the window

### PR DESCRIPTION
**Pull request by KungFuJesus**
Evidently we weren't hitting this code path very often in our tests?
If we reach here, there are some circumstances where we never ran any
of the bytes through the checksum. We can easily check for this with a
comparison if the initial checksum bytes.
    
This should fix issue #1600, and possibly #1565 as well.

### CODE REVIEW

**Comment by KungFuJesus** [comment]
OK, I think this will work.  It's maybe a bit hackier than I'd like but I'm pretty sure correctness isn't an issue.  This should catch the corner case we were seeing.  The main difference is I think there were unconditional checksums being applied in earlier versions in a lot of places.  I don't see a clean delineation where we clearly aren't applying the checksum but should be, though.

**Comment by KungFuJesus** [comment]
Hmm, I have half a mind to write the buffer being fed through that library's test suite and put it into ours.

**Comment by KungFuJesus** [comment]
Not sure what's up with the clang reduced memory failure.

Ahh, it does look like it reduces the window size to 8k, so perhaps this is making us hit a different corner case.  It also has dawned on me that the checksum will always be the initial CRC32 value until the final fold so that bit might need to be reworked.

**Comment by KungFuJesus** [comment]
I reproduced the reduced mem failure, I'll look into it tonight.

**Comment by KungFuJesus** [comment]
Ok I think I got it.  Still a little hackier than I'd like (I'd be much more thrilled about catching the set of conditions that paints us in this corner) but barring better suggestions or proof that this is not catching all of the corner cases, I say it's ready to merge.

File: inflate.c
Comment (mtl1979) [inline inflate.c]: Adding this empty line seems unnecessary... Mixing white-space and code changes is frowned upon...

File: arch/x86/crc32_fold_pclmulqdq_tpl.h
Comment (mtl1979) [inline arch/x86/crc32_fold_pclmulqdq_tpl.h]: Having a comment to explain the reason for why this is set to `1` is a good thing, but for the purpose of reviewing, it is still unclear why the value is exactly `1` and not some "magic constant" like elsewhere.

It is unclear what is `later`... So I would expand it to say exactly which function checks that this value is `1` and not any other value.

Comment (KungFuJesus) [reply arch/x86/crc32_fold_pclmulqdq_tpl.h]: How's that?

Comment (nmoinvaz) [reply arch/x86/crc32_fold_pclmulqdq_tpl.h]: Where is the `crc->value` actually being checked?

File: inflate.c
Comment (nmoinvaz) [inline inflate.c]: It would be great if there was a way we could clean up this if statement so it doesn't break like this with the `#define`.

File: arch/x86/crc32_fold_pclmulqdq_tpl.h
Comment (nmoinvaz) [reply arch/x86/crc32_fold_pclmulqdq_tpl.h]: Oh I see, it is checked as `crc_fold.value == CRC32_INITIAL_VALUE` below.

File: inflate.c
Comment (nmoinvaz) [inline inflate.c]:
```suggestion
                    } else {
                       bool final_chksum = (!state->flags && strm->adler == ADLER32_INITIAL_VALUE);
#ifdef GUNZIP
                       final_chksum |= (state->flags && state->crc_fold.value == CRC32_INITIAL_VALUE));
#endif
                        /* In the event we reach this block and the checksums are at the initial values, run a checksum
                         * over the whole stream, as one of two things are true:
                         *  1.) We never ran any of the bytes through the checksum (testing has found this happens)
                         *  2.) We have a checksum that does wrap perfectly, in which case, we should end up still
                         *      correct, anyway */
                        if (final_chksum)
                            inf_chksum(strm, put, strm->total_out);
                    }
```

File: arch/x86/crc32_fold_pclmulqdq_tpl.h
Comment (nmoinvaz) [reply arch/x86/crc32_fold_pclmulqdq_tpl.h]: Does this assignment need to be wrapped in `#ifdef GUNZIP` too?

Comment (KungFuJesus) [reply arch/x86/crc32_fold_pclmulqdq_tpl.h]: I don't think so, it's a function only actually called when using the crc32 folding routines, which would be enabled anyway with GUNZIP being defined.  Technically these functions are used in the generic crc32 exported function as well, but the extra assignment is hardly consequential, I think.

File: inflate.c
Comment (KungFuJesus) [reply inflate.c]: It does require C99 to do this but I don't think that's a huge ask.  We probably already require it.

**Review by Dead2 — APPROVED** [review]
LGTM

**Comment by Dead2** [comment]
Wish we did have a CI test for this, so we could make sure that this is not a problem for all the other alternative code-paths as well. I don't think any other arch based their code on `crc32_fold_pclmulqdq` etc and inherited the problem, did they?

**Comment by KungFuJesus** [comment]
The problem isn't specific to crcs, it's specific to inflate. There's a potential encoding through the compression state machine that arrives at the final check instruction without the current window getting run through the checksum routines.

This seems to catch those corner cases, as when it happens the checksum is the initial unaltered checksum, be that Adler or crc. 

As far as something that catches this, I think integrating libgit2's tests into our CI wouldn't be a bad approach.

**Review by mtl1979 — APPROVED** [review]
LGTM... We can add CI test in separate pull request.

File: inflate.c
Comment (nmoinvaz) [inline inflate.c]: @Dead2 we okay using this include? First in the library. I had mentioned it before.

**Comment by nmoinvaz** [comment]
I am testing using `minideflate` and comparing to `libgit2`. The `put` variable sent to `inf_chksum` is different.

image

`minideflate` using a next position in write buffer for each call to `inflate()`, where as `libgit2` starts from scratch from the buffer, so `put` is evaluated correctly.

Here is my test data:
[000000A2CD78EBA0-inflate.gz](https://github.com/zlib-ng/zlib-ng/files/13433188/000000A2CD78EBA0-inflate.gz)
```bash
minideflate -d -k -s 4 -r 25 -w 15 -t 8192 000000A2CD78EBA0-inflate.gz
```

And change `minideflate`:
```c
            err = PREFIX(inflate)(&d_stream, flush);
            if (err == Z_BUF_ERROR && read_buf_size != BUFSIZE) {
                err = Z_OK;
            }
```

So essentially, it only works because in `libgit2` they are passing the same buffer pointer of the previous `inflate()` to the the next call to `inflate()`. I still don't think we should commit this until we add a unit test.

**Comment by nmoinvaz** [comment]
This also seems to work although it can probably be better some how:
```diff
  inf_leave:
    RESTORE();
    if (INFLATE_NEED_UPDATEWINDOW(strm) &&
            (state->wsize || (out != strm->avail_out && state->mode < BAD &&
                 (state->mode < CHECK || flush != Z_FINISH)))) {
        /* update sliding window with respective checksum if not in "raw" mode */
        if (updatewindow(strm, strm->next_out, out - strm->avail_out, state->wrap & 4)) {
            state->mode = MEM;
            return Z_MEM_ERROR;
        }
-   }
+   } else if (!state->wsize && flush == Z_FINISH) {
+       inf_chksum(strm, put - (out - strm->avail_out), (out - strm->avail_out));
+   }
```
Vars: `state->wsize = 0, flush = Z_FINISH, state->mode = CHECK`

In this use case `Z_FINISH` is being used which according to the docs:
image

Normally, `inflate` might already have a window if it had been previously called with another flush type. But in `libgit2` case it is very short length and calling it with `Z_FINISH`, a window doesn't get created or used.

**Comment by KungFuJesus** [comment]
> This also seems to work although it can probably be better some how:
> 
> ```diff
>   inf_leave:
>     RESTORE();
>     if (INFLATE_NEED_UPDATEWINDOW(strm) &&
>             (state->wsize || (out != strm->avail_out && state->mode < BAD &&
>                  (state->mode < CHECK || flush != Z_FINISH)))) {
>         /* update sliding window with respective checksum if not in "raw" mode */
>         if (updatewindow(strm, strm->next_out, out - strm->avail_out, state->wrap & 4)) {
>             state->mode = MEM;
>             return Z_MEM_ERROR;
>         }
> -   }
> +   } else if (!state->wsize && flush == Z_FINISH) {
> +       inf_chksum(strm, put - (out - strm->avail_out), (out - strm->avail_out));
> +   }
> ```
> 
> Vars: `state->wsize = 0, flush = Z_FINISH, state->mode = CHECK`
> 
> In this use case `Z_FINISH` is being used which according to the docs: image
> 
> Normally, `inflate` might already have a window if it had been previously called with another flush type. But in `libgit2` case it is very short length and calling it with `Z_FINISH`, a window doesn't get created or used.

That solution is a lot cleaner assuming it captures every instance of this.  I was hoping to find a solution more like that but couldn't bring myself to find the right spot and conditions to inject that extra forced checksum.  So the right conditions are specifically if the user has specified Z_FINISH, bypassing the window creation entirely and effectively it's moved to the end stage "check" state

**Comment by KungFuJesus** [comment]
I like @nmoinvaz's solution much better, way less of a hack, way fewer lines, fewer comparisons. @nmoinvaz can you file a PR with your minideflate changes and that simple reproducer as a test case?

File: inflate.c
Comment (KungFuJesus) [reply inflate.c]: Don't think we need this, I went with your latter suggested approach which rendered it unnecessary.

**Comment by nmoinvaz** [comment]
Yes, I can create a PR with my `minideflate` use case. I looked putting my suggested logic in `CHECK` mode instead, but `CHECK` mode bails early in `NEEDSBITS` with `Z_BUF_ERROR`. So I think at the bottom of the function like I originally suggested is probably the best place for the logic.

File: inflate.c
Comment (nmoinvaz) [inline inflate.c]: I am wondering if it should only be done if also `&& err == Z_BUF_ERROR` -- that is, only when `NEEDBITS()` bails early. I'm not 100% sure.

Comment (KungFuJesus) [reply inflate.c]: Are there conditions where we'd hit this and err != Z_BUF_ERROR?  I think the spirit of the original code had a checksum here a bit less conditionally (the conditions that applied before were basically if (out)).

Comment (nmoinvaz) [reply inflate.c]: The reason why I suggest it is: because the only test that fails is when CHECK bails early in NEEDBITS. When CHECK doesn't bail early I assume it will do the chksum in that case statement...?

Comment (mtl1979) [reply inflate.c]: I think adding check `&& err == Z_BUF_ERROR` might be warranted, but we will need another test for that to be 100% sure... Returning `Z_BUF_ERROR` tells me the stream is truncated in some way and that itself should be enough to not silently ignore the case even if that means there is another bug somewhere else.

File: inflate.c
Comment (mtl1979) [inline inflate.c]: Using same expression twice hurts readability... even if using temporary variable means adding one more line, this is still quite short delta.

**Review by mtl1979 — CHANGES_REQUESTED** [review]
Needs further testing...

File: inflate.c
Comment (nmoinvaz) [reply inflate.c]: To check the other scenario where it doesn't bail early, just don't send `-r 25` to minideflate from my PR.

Comment (KungFuJesus) [reply inflate.c]: Sorry, was doing some traveling for the holiday.  On it, now.

File: inflate.c
Comment (KungFuJesus) [reply inflate.c]: I'm not entirely sure what you mean, none of the conditions being checked here overlap.  It just so happens that NEED_UPDATEWINDOW and NEED_CHECKSUM are the same in most cases, but they wouldn't necessarily have to be.  And inf_chksum does something very different from updatewindow.

File: inflate.c
Comment (KungFuJesus) [reply inflate.c]: k, pushed update.

File: inflate.c
Comment (KungFuJesus) [reply inflate.c]: Are you referring to the pointer arithmetic?  I guess that can be baked out.  One second.

Comment (KungFuJesus) [reply inflate.c]: I think that's what you mean, let me know if not.  Resolving for now.

File: inflate.c
Comment (nmoinvaz) [inline inflate.c]: I think maybe you meant to do snakecase instead of camelcase.

Comment (KungFuJesus) [reply inflate.c]: ah, sorry camel is my personal preference that's hard to break. Fixing now.

Comment (nmoinvaz) [reply inflate.c]: I use it all the time too because like 70% of what I do is in JavaScript these days.

File: inflate.c
Comment (nmoinvaz) [inline inflate.c]: Shoot, I just realized checking for `ret == Z_BUF_ERROR` at this place will not work, because `PULLBYTE` (called from `NEEDBITS`) only sets `ret = Z_BUF_ERROR;` in `infback.c`, which is weird. The `PULLBYTE` defined in `inflate.c` does not do that.

Comment (nmoinvaz) [reply inflate.c]: I wonder now if the check should be moved into `ret = Z_BUF_ERROR` below. 
```diff
if (((in == 0 && out == 0) || flush == Z_FINISH) && ret == Z_OK) {
+    /* when no sliding window is used, hash the output bytes when CHECK state does not */
+    if (INFLATE_NEED_CHECKSUM(strm) && !state->wsize && flush == Z_FINISH) {
+        inf_chksum(strm, put - check_bytes, check_bytes);
+    }
    ret = Z_BUF_ERROR;
}
```

Comment (KungFuJesus) [reply inflate.c]: Yeah, I think it might make more sense there.

File: inflate.c
Comment (nmoinvaz) [reply inflate.c]: FYI, I added the other check to the test cases in #1603.

**Comment by nmoinvaz** [comment]
LGTM, now we just need to rebase after #1603 is merged.

File: inflate.c
Comment (mtl1979) [reply inflate.c]: I think having both tests is better if just to catch potential regressions... We can't just test existing failures (before fixing them)...

File: inflate.c
Comment (mtl1979) [reply inflate.c]: I meant the `(out - strm->avail_out)` part, which you did fix.... I know compiler is smart enough to do duplicate code elimination even on lower optimisation levels, but we really shouldn't rely on it.

**Comment by Dead2** [comment]
Rebase please 👍

**Review by Dead2 — APPROVED** [review]
I like this elegant solution 👍

**Comment by KungFuJesus** [comment]
Guessing we can probably ignore that failure to upload issue?

**Comment by KungFuJesus** [comment]
Rerolling the dice anyway for that green check mark.

**Comment by Dead2** [comment]
Looking good, all tests passing. I will probably roll out a hotfix release with this within a day or two, unless something crops up.

**Review by mtl1979 — APPROVED** [review]
LGTM

**Pull request merged**
